### PR TITLE
INT-3878: MQTT Application Events (inbound)

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3878

Publish `ApplicationEvent`s to report inbound channel adapter connection
status.

### CODE REVIEW

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java [hunk 1]
```diff
@@ -103,6 +112,24 @@ public void setCompletionTimeout(int completionTimeout) {
 		this.completionTimeout = completionTimeout;
 	}
 
+	/**
+	 * The time (ms) to wait between reconnection attempts. Default
+	 * {@value #DEFAULT_RECOVERY_INTERVAL}.
+	 * @param recoveryInterval the interval.
+	 * @since 4.2.2
+	 */
+	public void setRecoveryInterval(int recoveryInterval) {
```
Comment (artembilan) [inline spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java]: Shouldn't we consider to use `BackOff` abstraction as well?

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java [hunk 2]
```diff
@@ -186,18 +213,21 @@ private void connectAndSubscribe() throws MqttException {
 		MqttConnectOptions connectionOptions = this.clientFactory.getConnectionOptions();
 		Assert.state(this.getUrl() != null || connectionOptions.getServerURIs() != null,
 				"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
-		this.client = this.clientFactory.getAsyncClientInstance(this.getUrl(), this.getClientId());
+		this.client = this.clientFactory.getAsyncClientInstance(this.getUrl(), getClientId());
```
Comment (artembilan) [inline spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java]: `this.` for the `getUrl()` can be removed as well :smile:

File: spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java (hunk 1 above)
Comment (garyrussell) [inline spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java]: Perhaps, but in 4.3 only (new JIRA).

File: src/reference/asciidoc/mqtt.adoc [hunk 3]
```diff
@@ -80,7 +80,14 @@ The `DefaultPahoMessageConverter` can be configured to return the raw `byte[]` i
 NOTE: Starting with _version 4.1_ the url can be omitted and, instead, the server URIs can be provided in the `serverURIs` property of the `DefaultMqttPahoClientFactory`.
 This enables, for example, connection to a highly available (HA) cluster.
 
-
+Starting with _version 4.2.2_, an `MqttSubscribedEvent` is published when the adapter successfully subscribes to the
+topic(s).
+`MqttConnnectionFailedEvent` s are published when the connection/subscription fails.
```
Comment (artembilan) [inline src/reference/asciidoc/mqtt.adoc]: Triple `n`. Will be fixed on merge.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/b486247105c05284c9187a22697262e47a1c453d

**Pull request closed without merging**